[SC-261] Naprawić, aby przechodząc przez ekspedycję nie można było się cofać w pytaniach

ajablonsk1 opened the pull request:
Witam, pare ważnych kwestii:

1. Usuąłem pole expirationDate (czy jakoś tak) z tasków (bo mamy to w wymaganiach), więc np przy tworzeniu graph taska i innych w jsonie nie trzeba podawać expirationDate, ale to chyba jest zalatwione, bo zwracam wam obiekt okrojony bez tego to chyba git.

2. Przechodzenie przez ekspedycje powinno działać już mniej więcej dobrze, przeszedłem sobie teraz przez jedną w taki sposób:

```
POST localhost:8080/api/task/graph/result/start

BODY: 
{
    "graphTaskId": 1
}
```


```
GET localhost:8080/api/question/next?graphTaskId=1
RESPONSE:
[
    {
        "id": 2,
        "type": "MULTIPLE_CHOICE",
        "hint": "Kable",
        "difficulty": "EASY",
        "points": 10.0
    },
    {
        "id": 3,
        "type": "SINGLE_CHOICE",
        "hint": "Manchester",
        "difficulty": "MEDIUM",
        "points": 20.0
    },
    {
        "id": 4,
        "type": "OPENED",
        "hint": "?",
        "difficulty": "HARD",
        "points": 30.0
    }
]
```
```
localhost:8080/api/question/info?questionId=2&graphTaskId=1
RESPONSE:
{
    "id": 2,
    "content": "Które urządzenia można połączyć ze sobą skrętką “prostą”?",
    "points": 10.0,
    "timeRemaining": 595269,
    "optionInfos": [
        {
            "id": 1,
            "content": "hub z routerem"
        },
        {
            "id": 2,
            "content": "komputer z komputerem"
        },
        {
            "id": 3,
            "content": "switch z routerem"
        },
        {
            "id": 4,
            "content": "hub ze switchem"
        }
    ]
}
```
```
http://localhost:8080/api/task/graph/result/answer/add
BODY
{
    "graphTaskId": 1,
    "questionId": 2,
    "answerForm": {
        "optionIds": [1, 2]
    }
}
RES
414896
```
I dalej zrobiłem znowy /next i pobrało mi kolejne pytania. Oczywiście sprawdziłem czy mogę sie wracać tak:

1. Gdy chciałem wrócić stąd localhost:8080/api/question/info?questionId=2&graphTaskId=1 do localhost:8080/api/question/next?graphTaskId=1 to mam error ze zły status jest i nie można .
```
{
    "status": "BAD_REQUEST",
    "timestamp": "17-09-2022 01:17:33",
    "message": "GraphTaskResult with id 5 should has CHOOSE status",
    "debugMessage": null
}
```
2. Gdy chciałem po http://localhost:8080/api/task/graph/result/answer/add wrócić do localhost:8080/api/question/info?questionId=2&graphTaskId=1, to mam, że na to pytanie już odpowiedział.
```
{
    "status": "BAD_REQUEST",
    "timestamp": "17-09-2022 01:18:17",
    "message": "Question was already answered!",
    "debugMessage": null
}
```
3. Gdy próbuje wrócić do localhost:8080/api/task/graph/result/start to mam
```
{
    "status": "CONFLICT",
    "timestamp": "17-09-2022 01:16:08",
    "message": "GraphTaskResult for GraphTask with id 1 and user murbanska@student.agh.edu.pl already exists",
    "debugMessage": null
}
```

Sprawdźcie proszę czy działa jak powinno i jak by była potrzeba to w message mogę wpisać coś po polsku żebyście łapali i wyświetlali.

ajablonsk1 committed:
[SC-261] Looks like it's working :D

kwdrt commented:
Widzę pewien problem - brakuje wybierania pytań kiedy patrzymy na następne pytania - ktoś może odczytać treść dowolnego z nich zanim udzieli odpowiedzi.

Jeszcze nie jestem pewien odnośnie rozbicia tego na /next i /info, jeśli załatwimy to jednym endpointem to możemy w zależności od aktualnego stanu wyświetlać tylko pytania do wyboru, albo tylko konkretne wybrane pytanie w zależności od stanu danego podejścia na serwerze - wtedy nie mamy sytuacji że ktoś sobie sprawdzi wszystkie pytania "do przodu" i wybierze to najłatwiejsze/co umie, bo back dostarczy mu tylko "prawidłowe" dane.


/add (w obecniej formie) też wtedy nie potrzebowałby questionId, bo w sumie nie ma po co - wystarczy mu graphTask żeby stwierdzić jakie jest aktualne pytanie ze stanu graphTaska


Zastanowiłbym się nad zamianą endpointów (`/info`, `/next`, `/add`) na:

- `/setState`, który by przyjmował:
```js
{
  action: 'CHOOSE' | 'ANSWER'
  answerForm: null | ten co teraz
  answerPick: null | id wybranego pytania
}
```
i blokował odpowiednio albo"funkcję" 'CHOOSE' albo 'ANSWER' w zależności od tego czy jest na etapie udzielania odpowiedzi do konkretnego pytania, czy już to zrobił, a także czy pytanie jest w następnych
mógłby też zastąpić wtedy /add. W przypadku udzielania odpowiedzi powinien móc wybrać pytanie ze stanu na backu, bez podawania `questionId`


- `/getInfo`, który nie przyjmowałby nic oprócz id graphTaska, i tak samo zwracałby:
```js
{
  responseType: 'CHOOSE' | 'ANSWER'
  questionsToPick: null | lista ta co teraz
  questionDetails: null | detale pytania tak jak teraz
}
```

Gdzie 'CHOOSE' działałby jak obecny /next, 'ANSWER' działałby jak obecne /info, z odpowiednimi guardami.

Oczywiście w tych dwóch endpointach byłaby tylko jedna opcja naraz, pozostałe byłyby nullowane

kwdrt commented:
Jeszcze znalazłem przy okazji miejsce w którym "wycieka" nam lista odpowiedzi:

![firefox_gka3Wx0fB0](https://user-images.githubusercontent.com/56168780/190856167-b6ee54c9-6846-4dbf-b724-756922103cec.png)
![firefox_wvndIyxUH2](https://user-images.githubusercontent.com/56168780/190856174-c5803859-9330-4da4-9d92-19a4f7b3c817.png)

Pole `activity` wyrzuca nam wszystkie informacje o aktywności z bazy, włącznie z listą pytań wraz z odpowiedziami

ajablonsk1 commented:
> Widzę pewien problem - brakuje wybierania pytań kiedy patrzymy na następne pytania - ktoś może odczytać treść dowolnego z nich zanim udzieli odpowiedzi.
> 
> Jeszcze nie jestem pewien odnośnie rozbicia tego na /next i /info, jeśli załatwimy to jednym endpointem to możemy w zależności od aktualnego stanu wyświetlać tylko pytania do wyboru, albo tylko konkretne wybrane pytanie w zależności od stanu danego podejścia na serwerze - wtedy nie mamy sytuacji że ktoś sobie sprawdzi wszystkie pytania "do przodu" i wybierze to najłatwiejsze/co umie, bo back dostarczy mu tylko "prawidłowe" dane.
> 
> /add (w obecniej formie) też wtedy nie potrzebowałby questionId, bo w sumie nie ma po co - wystarczy mu graphTask żeby stwierdzić jakie jest aktualne pytanie ze stanu graphTaska
> 
> Zastanowiłbym się nad zamianą endpointów (`/info`, `/next`, `/add`) na:
> 
> * `/setState`, który by przyjmował:
> 
> ```js
> {
>   action: 'CHOOSE' | 'ANSWER'
>   answerForm: null | ten co teraz
>   answerPick: null | id wybranego pytania
> }
> ```
> 
> i blokował odpowiednio albo"funkcję" 'CHOOSE' albo 'ANSWER' w zależności od tego czy jest na etapie udzielania odpowiedzi do konkretnego pytania, czy już to zrobił, a także czy pytanie jest w następnych mógłby też zastąpić wtedy /add. W przypadku udzielania odpowiedzi powinien móc wybrać pytanie ze stanu na backu, bez podawania `questionId`
> 
> * `/getInfo`, który nie przyjmowałby nic oprócz id graphTaska, i tak samo zwracałby:
> 
> ```js
> {
>   responseType: 'CHOOSE' | 'ANSWER'
>   questionsToPick: null | lista ta co teraz
>   questionDetails: null | detale pytania tak jak teraz
> }
> ```
> 
> Gdzie 'CHOOSE' działałby jak obecny /next, 'ANSWER' działałby jak obecne /info, z odpowiednimi guardami.
> 
> Oczywiście w tych dwóch endpointach byłaby tylko jedna opcja naraz, pozostałe byłyby nullowane

Dobra zmienie 1:1 jak chcesz, myślałem, że to co zrobiłem będzie działać przy nieznacznym zmianie backendu, ale no nie będę już kombinować xd

ajablonsk1 commented:
> Jeszcze znalazłem przy okazji miejsce w którym "wycieka" nam lista odpowiedzi:
> 
> ![firefox_gka3Wx0fB0](https://user-images.githubusercontent.com/56168780/190856167-b6ee54c9-6846-4dbf-b724-756922103cec.png) ![firefox_wvndIyxUH2](https://user-images.githubusercontent.com/56168780/190856174-c5803859-9330-4da4-9d92-19a4f7b3c817.png)
> 
> Pole `activity` wyrzuca nam wszystkie informacje o aktywności z bazy, włącznie z listą pytań wraz z odpowiedziami

Używacie tego endpointa? jak nie to go wywale, a jak tak go zrobie jakieś sendowne dto więc napisz co potrzebujecie z niego

Commit:
Changed graph task flow

ajablonsk1 committed:
[SC-261] Merge branch 'develop' into SC-261-naprawic-aby-przechodzac-przez-ekspedycje-nie-mozna-bylo-sie-cofac-w-pytaniach

# Conflicts:
#	api/src/main/java/com/example/api/service/activity/task/GraphTaskService.java
#	api/src/main/java/com/example/api/service/validator/activity/ActivityValidator.java

ajablonsk1 committed:
[SC-261] Merge remote-tracking branch 'origin/SC-261-naprawic-aby-przechodzac-przez-ekspedycje-nie-mozna-bylo-sie-cofac-w-pytaniach' into SC-261-naprawic-aby-przechodzac-przez-ekspedycje-nie-mozna-bylo-sie-cofac-w-pytaniach

ajablonsk1 committed:
[SC-261] Works

ajablonsk1 committed:
[SC-261] Names correction

ajablonsk1 commented:
Dobra powinno działać tak jak @kwdrt chciał :). 

Startujemy ekspedycje:
```
POST localhost:8080/api/task/graph/result/start
BODY:
 {
    "graphTaskId": 1
}
```
Dostajemy info w zależnosci od stanu ekspedycji:
```
GET localhost:8080/api/question/get-info?graphTaskId=1
{
    "status": "CHOOSE",
    "timeRemaining": 718060,
    "questions": [
        {
            "id": 2,
            "type": "MULTIPLE_CHOICE",
            "hint": "Kable",
            "difficulty": "EASY",
            "points": 10.0
        },
        {
            "id": 3,
            "type": "SINGLE_CHOICE",
            "hint": "Manchester",
            "difficulty": "MEDIUM",
            "points": 20.0
        },
        {
            "id": 4,
            "type": "OPENED",
            "hint": "?",
            "difficulty": "HARD",
            "points": 30.0
        }
    ],
    "questionDetails": null
}
OR 
{
    "status": "ANSWER",
    "timeRemaining": 680704,
    "questions": null,
    "questionDetails": {
        "questionId": 2,
        "content": "Które urządzenia można połączyć ze sobą skrętką “prostą”?",
        "points": 10.0,
        "options": [
            {
                "id": 1,
                "content": "hub z routerem"
            },
            {
                "id": 2,
                "content": "komputer z komputerem"
            },
            {
                "id": 3,
                "content": "switch z routerem"
            },
            {
                "id": 4,
                "content": "hub ze switchem"
            }
        ]
    }
}
```
Wybieramy pytanie lub odpowiadamy na nie:
```
POST localhost:8080/api/question/action
BODY:
{
    "status": "CHOOSE",
    "graphTaskId": 1,
    "questionId": 2
}
OR 
{
    "status": "ANSWER",
    "graphTaskId": 1,
    "answerForm": {
        "optionIds": [1]
    }
}
```
Co do answer form to tam jest albo lista optionIds albo openAnswer. Dla single choice dajecie jeden id do listy, dla multiple tyle ile wybierze student. Jeżeli coś nie będzie grać to wywali errora (np dla single choice jak dwa id dacie). Dla opened oczywiście dajecie openedAnswer jako stringa. 

POST localhost:8080/api/question/action  - zwraca nam też timeRemaining

EDIT: Zmieniłem jeszcze endpointa GET /api/task/graph/result i zwraca on teraz tylko id zamiast całego obiektu

ajablonsk1 committed:
[SC-261] Changed one endpoint to not return the whole object

kwdrt commented:
endpoint setState nie blokuje się po przekroczeniu czasu:
![obraz](https://user-images.githubusercontent.com/56168780/191278247-c0ad973b-c604-4ed6-90c5-b491b1d7acd5.png)

Widzę że dostaję 400 (bez konkretnego błędu) kiedy próbuję zrobić ANSWER na przekroczonym czasie (ale nie CHOOSE) - tutaj dopiero wchodzi guard? Jeśli tak to w sumie bez problemu

ajablonsk1 commented:
> endpoint setState nie blokuje się po przekroczeniu czasu: ![obraz](https://user-images.githubusercontent.com/56168780/191278247-c0ad973b-c604-4ed6-90c5-b491b1d7acd5.png)
> 
> Widzę że dostaję 400 (bez konkretnego błędu) kiedy próbuję zrobić ANSWER na przekroczonym czasie (ale nie CHOOSE) - tutaj dopiero wchodzi guard? Jeśli tak to w sumie bez problemu

Nie wysylam bledu jak czas jest przekroczony, ale tez nie zapisuje odpowiedzi. Moge zmienic zeby error sie wysylal, aktualnie w tamtej funkcji jest if ze jak czas jest ponizej zera to zwroc czas i tyle, nic wiecej nie robi.

Ok, widzę jeszcze że się ustawia pozostały czas na CHOOSE, ale to raczej nie jest problem, myślę że to jest git

ajablonsk1 committed:
[SC-261] Added actual points to get info

ajablonsk1 commented:
Dobra dodałem to wszystko, jutro mnie na kompoe nie będzie raczej więc jak ktoś zrobi review i będzie git to merge jak nie to @barto14753 będzie musiał naprawić

ajablonsk1 committed:
Merge branch 'develop' into SC-261-naprawic-aby-przechodzac-przez-ekspedycje-nie-mozna-bylo-sie-cofac-w-pytaniach

kwdrt committed:
fixed getting result after endpoint change

kwdrt commented:
Przeszedłem przez flow znowu po zmianach - wygląda dobrze z liczeniem punktów i nie udało mi się spowodować żadnych niewłaściwych zachowań podczas testów, dodałem tylko fix na resultId na froncie po zmianie endpointa

kwdrt reviewed (approved):
LGTM 👍

kwdrt commented:
Obecnie przechodzenie przez ekspedycję na froncie nie działa poza endpointami przez zmianę flow - będę nad tym pracował
